#                                                       **Fresco Retailer Product Return Prediction** 

In [2]:
import pandas as pd
import numpy as np
import plotly.express as px

## Upload data 

In [3]:
df = pd.read_excel('/Users/mohammedmahmood/Desktop/Data projects/Data science/Fresco Retailer Product Return Prediction/Fresco_Retailerr.xlsx')

In [4]:
df.sample(5)

transaction_id     Month product_category Product_Subcategory  Qty  \
7844      74750830693       May         Clothing                Mens    3   
18312     32207223462  December         Footwear                Mens    5   
8280      98337096209       May         Clothing                Mens    5   
16117      8903414497  November         Footwear                Kids    3   
18882     63981520035  December         Footwear                Mens    2   

       Rate  Amount      Tax  Delivery_chrgs     Payment_mode  ... Reviews  \
7844    614    1842  193.410             NaN  Mobile Payments  ...     1.0   
18312   525    2625  275.625           140.0       Debit Card  ...     1.0   
8280   1339    6695  702.975           250.0      Credit Card  ...     5.0   
16117  1297    3891  408.555             NaN      Credit Card  ...     1.0   
18882   392     784   82.320            50.0  Mobile Payments  ...     5.0   

       CustID   Income            city Return        DOB Gender  \
7844     1894  74822.0  Vishakhapatnam  False 1976-12-15      M   
18312   10812  83569.0       Bangalore  False 1965-06-28      F   
8280    12084  46126.0  Vishakhapatnam  False 1976-07-28      M   
16117    4175  43806.0       Bangalore  False 1967-12-27      F   
18882    5095  27795.0            Pune  False 1964-10-23      F   

      Marital_status  Education_Level Profession_Type  
7844               0           Others        Salaried  
18312              1            Basic          Others  
8280               1              PhD          Others  
16117              1       Graduation        Salaried  
18882              0           Master   Self Employed  

[5 rows x 21 columns]

In [5]:
df.shape

(19597, 21)

In [6]:
df.columns

Index(['transaction_id', 'Month', 'product_category', 'Product_Subcategory',
       'Qty', 'Rate', 'Amount', 'Tax', 'Delivery_chrgs', 'Payment_mode',
       'Store_type', 'Reviews', 'CustID', 'Income', 'city', 'Return', 'DOB',
       'Gender', 'Marital_status', 'Education_Level', 'Profession_Type'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19597 entries, 0 to 19596
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   transaction_id       19597 non-null  int64         
 1   Month                19597 non-null  object        
 2   product_category     19597 non-null  object        
 3   Product_Subcategory  19597 non-null  object        
 4   Qty                  19597 non-null  int64         
 5   Rate                 19597 non-null  int64         
 6   Amount               19597 non-null  int64         
 7   Tax                  19597 non-null  float64       
 8   Delivery_chrgs       10707 non-null  float64       
 9   Payment_mode         19597 non-null  object        
 10  Store_type           19597 non-null  object        
 11  Reviews              19546 non-null  float64       
 12  CustID               19597 non-null  int64         
 13  Income               19554 non-

## All data types right

## nulls

In [8]:
df.isna().sum() 

transaction_id            0
Month                     0
product_category          0
Product_Subcategory       0
Qty                       0
Rate                      0
Amount                    0
Tax                       0
Delivery_chrgs         8890
Payment_mode              0
Store_type                0
Reviews                  51
CustID                    0
Income                   43
city                      0
Return                    0
DOB                       0
Gender                    0
Marital_status            0
Education_Level           0
Profession_Type           0
dtype: int64

## duplicated

In [9]:
df.duplicated().sum()

np.int64(0)

In [10]:
df.columns

Index(['transaction_id', 'Month', 'product_category', 'Product_Subcategory',
       'Qty', 'Rate', 'Amount', 'Tax', 'Delivery_chrgs', 'Payment_mode',
       'Store_type', 'Reviews', 'CustID', 'Income', 'city', 'Return', 'DOB',
       'Gender', 'Marital_status', 'Education_Level', 'Profession_Type'],
      dtype='object')

## Univariate analysis

### **In this step i will go through each column and fix problems in and do Eda for features to understand**

In [11]:
# unneccery in prediction 
df.drop("transaction_id", axis= 1, inplace= True)

In [12]:
df.duplicated().sum()

np.int64(0)

## Month

In [13]:
df.Month.unique()

array(['January', 'February', 'March', 'April', 'May', 'June', 'July',
       'August', 'September', 'October', 'November', 'December'],
      dtype=object)

## product_category

In [14]:
df.product_category.unique()

array(['Books', 'Clothing', 'Home and kitchen', 'Footwear', 'Bags',
       'Electronics'], dtype=object)

In [15]:
df.product_category.value_counts()

product_category
Footwear            5027
Clothing            4286
Home and kitchen    3575
Bags                2693
Books               2426
Electronics         1590
Name: count, dtype: int64

## Product_Subcategory

In [16]:
df.Product_Subcategory.unique()

array(['Fiction', 'Women', 'Bath', 'Mens', 'Kitchen', 'Furnishing',
       'Mobiles', 'DIY', 'Kids', 'Non-Fiction', 'Audio and video',
       'Computers', 'Cameras', 'Tools', 'Academic', 'Personal Appliances'],
      dtype=object)

In [17]:
df.Product_Subcategory.value_counts()

Product_Subcategory
Women                  5494
Mens                   4879
Kids                   1633
Furnishing              951
Kitchen                 912
Bath                    859
Tools                   853
DIY                     642
Academic                615
Non-Fiction             596
Fiction                 573
Computers               331
Cameras                 327
Audio and video         326
Mobiles                 305
Personal Appliances     301
Name: count, dtype: int64

## Qty 

In [18]:
df.rename(columns={'Qty': 'Quantity'}, inplace=True)

In [19]:
df.Quantity.unique()

array([3, 5, 4, 2, 1])

In [20]:
df.Quantity.value_counts()

Quantity
5    4599
4    3941
1    3795
3    3637
2    3625
Name: count, dtype: int64

In [21]:
df.columns

Index(['Month', 'product_category', 'Product_Subcategory', 'Quantity', 'Rate',
       'Amount', 'Tax', 'Delivery_chrgs', 'Payment_mode', 'Store_type',
       'Reviews', 'CustID', 'Income', 'city', 'Return', 'DOB', 'Gender',
       'Marital_status', 'Education_Level', 'Profession_Type'],
      dtype='object')

## Rate

In [22]:
df.rename(columns={ 'Rate': 'Unit_Price'}, inplace=True)

In [23]:
df.Unit_Price.describe()

count    19597.000000
mean       777.154718
std        414.308709
min         70.000000
25%        418.000000
50%        775.000000
75%       1136.000000
max       1500.000000
Name: Unit_Price, dtype: float64

In [24]:
px.histogram(df, x= "Unit_Price" )

In [25]:
px.box(df, x= "Unit_Price" )

## Amount

In [26]:
df[["Quantity", "Unit_Price", "Amount", "Tax", "Delivery_chrgs" ] ].head(10)

Quantity  Unit_Price  Amount      Tax  Delivery_chrgs
0         3         359    1795  188.475            90.0
1         5        1129    5645  592.725           250.0
2         4        1327    6635  696.675             NaN
3         5        1325    6625  695.625             NaN
4         3        1237    6185  649.425             NaN
5         5         837     837   87.885            50.0
6         4         931    2793  293.265             NaN
7         2        1211    3633  381.465           200.0
8         1        1497    4491  471.555           200.0
9         5        1008    2016  211.680            90.0

## During data validation, we discovered inconsistencies in the Amount column. For example, in the first row:

Quantity = 3
Unit_Price = 359
Expected Amount = 3 × 359 = 1077

However, the existing value in the Amount column was 1795, which is incorrect.

To resolve this issue and ensure data accuracy, we recalculated the Amount BY multibly Quantity * Unit_Price

In [27]:
# drop column 
df.drop("Amount", axis= 1, inplace= True)

In [28]:
# recalculated column and named by Total_Price
df['Total_Price'] = df['Quantity'] * df['Unit_Price']

In [29]:
df.Total_Price.describe()

count    19597.000000
mean      2397.469664
std       1795.267407
min         70.000000
25%        948.000000
50%       1936.000000
75%       3534.000000
max       7500.000000
Name: Total_Price, dtype: float64

In [30]:
px.histogram(df, x= "Total_Price" ) 

In [31]:
px.box(df, x= "Total_Price" )

In [32]:
from datasist.structdata import detect_outliers 

outliers_indices = detect_outliers(df, 0 , ["Total_Price"]) 

len(outliers_indices)

62

## Tax

In [33]:
df.Tax.describe()

count    19597.000000
mean       241.722141
std        182.975767
min          7.350000
25%         96.075000
50%        191.730000
75%        351.540000
max        787.500000
Name: Tax, dtype: float64

In [34]:
px.histogram(df, x= "Tax" ) 

In [35]:
from datasist.structdata import detect_outliers 

outliers_indices = detect_outliers(df, 0 , ["Tax"]) 

len(outliers_indices)

265

In [36]:
df.columns

Index(['Month', 'product_category', 'Product_Subcategory', 'Quantity',
       'Unit_Price', 'Tax', 'Delivery_chrgs', 'Payment_mode', 'Store_type',
       'Reviews', 'CustID', 'Income', 'city', 'Return', 'DOB', 'Gender',
       'Marital_status', 'Education_Level', 'Profession_Type', 'Total_Price'],
      dtype='object')

## 

## Delivery_chrgs

In [37]:
df.drop("Delivery_chrgs", inplace= True, axis= 1)

## Payment_mode

In [38]:
df.Payment_mode.unique()

array(['Mobile Payments', 'Credit Card', 'Debit Card', 'Cash'],
      dtype=object)

In [39]:
df.Payment_mode.value_counts()

Payment_mode
Mobile Payments    8191
Credit Card        4683
Debit Card         4677
Cash               2046
Name: count, dtype: int64

## Store_type

In [40]:
df.Store_type.unique()

array(['e-Shop', 'TeleShop', 'MBR', 'Flagship store'], dtype=object)

In [41]:
df.Store_type.value_counts()

Store_type
e-Shop            8014
MBR               3886
TeleShop          3872
Flagship store    3825
Name: count, dtype: int64

## Reviews

In [42]:
df.Reviews.unique()

array([ 1.,  4.,  2.,  5.,  3., nan])

In [43]:
df['Reviews'].isnull().sum()

np.int64(51)

In [44]:
df = df.dropna(subset=['Reviews'])

In [ ]:
# remave nulll in column because it dosn't make sense to impute Reviews
df['Reviews'].isnull().sum()

np.int64(0)

In [46]:
df.Reviews.value_counts()

Reviews
5.0    9537
4.0    4696
1.0    2491
2.0    1542
3.0    1280
Name: count, dtype: int64

In [47]:
df.columns

Index(['Month', 'product_category', 'Product_Subcategory', 'Quantity',
       'Unit_Price', 'Tax', 'Payment_mode', 'Store_type', 'Reviews', 'CustID',
       'Income', 'city', 'Return', 'DOB', 'Gender', 'Marital_status',
       'Education_Level', 'Profession_Type', 'Total_Price'],
      dtype='object')

## CustID

In [48]:
df.drop("CustID", inplace= True, axis= 1)

## Income

In [49]:
df.Income.describe()

count     19504.000000
mean      70516.888074
std       37418.384648
min        7157.000000
25%       37950.000000
50%       69293.000000
75%       99534.000000
max      159984.000000
Name: Income, dtype: float64

In [50]:
df['Income'].isnull().sum()

np.int64(42)

In [51]:
df = df.dropna(subset=['Income'])

In [52]:
df['Income'].isnull().sum()

np.int64(0)

In [53]:
px.histogram(df, x= "Income" )

## Return

In [54]:
df.Return.unique()

array([ True, False])

In [109]:
df.Return.value_counts()

Return
0    17457
1     2047
Name: count, dtype: int64

In [ ]:
df["Return"] = df["Return"].map({False: 0, True: 1}) 

In [56]:
df.Return.value_counts()

Return
0    17457
1     2047
Name: count, dtype: int64

### **Target is immbalanced and wil handle it**

## DOB

In [57]:
df.DOB.unique() 

<DatetimeArray>
['1985-12-31 00:00:00', '1985-12-29 00:00:00', '1985-12-27 00:00:00',
 '1985-12-26 00:00:00', '1985-12-24 00:00:00', '1985-12-22 00:00:00',
 '1985-12-20 00:00:00', '1985-12-19 00:00:00', '1985-12-18 00:00:00',
 '1985-12-17 00:00:00',
 ...
 '1964-01-29 00:00:00', '1964-01-28 00:00:00', '1964-01-22 00:00:00',
 '1964-01-11 00:00:00', '1964-01-10 00:00:00', '1964-01-06 00:00:00',
 '1964-01-05 00:00:00', '1964-01-04 00:00:00', '1964-01-03 00:00:00',
 '1964-01-02 00:00:00']
Length: 3996, dtype: datetime64[ns]

In [58]:
df.drop("DOB", inplace= True, axis= 1 )
df.reset_index(inplace= True, drop= True)

## Gender

In [59]:
df.Gender.unique()

array(['M', 'F'], dtype=object)

In [60]:
df.Gender.value_counts()

Gender
M    12028
F     7476
Name: count, dtype: int64

## Marital_status

In [61]:
df.Marital_status.unique()

array([0, 1])

In [62]:
df.Marital_status.value_counts()

Marital_status
1    10316
0     9188
Name: count, dtype: int64

## Education_Level

In [63]:
df.Education_Level.unique()

array(['PhD', 'Master', 'Basic', 'Others', 'Graduation'], dtype=object)

## Profession_Type

In [64]:
df.Profession_Type.unique()

array(['Others', 'Salaried', 'Self Employed'], dtype=object)

In [65]:
df

Month  product_category Product_Subcategory  Quantity  Unit_Price  \
0       January             Books             Fiction         3         359   
1       January          Clothing               Women         5        1129   
2       January  Home and kitchen                Bath         4        1327   
3       January          Footwear                Mens         5        1325   
4       January          Clothing               Women         3        1237   
...         ...               ...                 ...       ...         ...   
19499  December       Electronics     Audio and video         5          79   
19500  December              Bags                Mens         1        1196   
19501  December          Footwear               Women         4         739   
19502  December       Electronics             Mobiles         1          71   
19503  December          Footwear                Mens         5        1268   

           Tax     Payment_mode Store_type  Reviews    Income       city  \
0      188.475  Mobile Payments     e-Shop      1.0   67501.0  Hyderabad   
1      592.725      Credit Card   TeleShop      1.0  102738.0  Bangalore   
2      696.675       Debit Card        MBR      1.0  104013.0    Kolkata   
3      695.625       Debit Card        MBR      4.0   92580.0  Hyderabad   
4      649.425       Debit Card        MBR      1.0  101763.0  New Delhi   
...        ...              ...        ...      ...       ...        ...   
19499    8.295  Mobile Payments     e-Shop      1.0   99618.0    Gurgaon   
19500  125.580  Mobile Payments   TeleShop      3.0   32595.0    Kolkata   
19501  310.380       Debit Card   TeleShop      1.0   82914.0  New Delhi   
19502    7.455  Mobile Payments     e-Shop      5.0   80254.0    Gurgaon   
19503  665.700      Credit Card     e-Shop      5.0   55376.0  Ahmedabad   

       Return Gender  Marital_status Education_Level Profession_Type  \
0           1      M               0             PhD          Others   
1           1      M               1          Master        Salaried   
2           1      M               1           Basic        Salaried   
3           1      M               1          Others          Others   
4           1      M               0           Basic          Others   
...       ...    ...             ...             ...             ...   
19499       1      M               0          Master   Self Employed   
19500       0      F               0          Master   Self Employed   
19501       0      F               1          Others   Self Employed   
19502       0      M               0          Others          Others   
19503       0      F               1      Graduation   Self Employed   

       Total_Price  
0             1077  
1             5645  
2             5308  
3             6625  
4             3711  
...            ...  
19499          395  
19500         1196  
19501         2956  
19502           71  
19503         6340  

[19504 rows x 17 columns]

In [66]:
df.columns

Index(['Month', 'product_category', 'Product_Subcategory', 'Quantity',
       'Unit_Price', 'Tax', 'Payment_mode', 'Store_type', 'Reviews', 'Income',
       'city', 'Return', 'Gender', 'Marital_status', 'Education_Level',
       'Profession_Type', 'Total_Price'],
      dtype='object')

# Feature Engineering 

## Tax Rate

In [67]:
df['tax_rate'] = df['Tax'] / df['Total_Price']

In [68]:
df.tax_rate.describe()

count    19504.000000
mean         0.103501
std          0.026438
min          0.021000
25%          0.105000
50%          0.105000
75%          0.105000
max          0.525000
Name: tax_rate, dtype: float64

## price_After_Tax

In [ ]:
df['price_After_Tax'] = (df['Total_Price'] + df['Tax']) 

# Feature Importance 

## check corr

In [70]:
df_Cat = df.select_dtypes(include= "number")
corr = df_Cat.corr()['Return'].sort_values(ascending=False)
print(corr) 

Return             1.000000
Quantity           0.264764
Total_Price        0.181667
price_After_Tax    0.168094
Income             0.096515
Tax                0.025561
Unit_Price         0.016190
Marital_status     0.004795
tax_rate          -0.165529
Reviews           -0.541104
Name: Return, dtype: float64


In [71]:
df_Cat = df.select_dtypes(include="object")

for col in df_Cat.columns :
    print(f"Column {col} has {df_Cat[col].nunique()} values")

Column Month has 12 values
Column product_category has 6 values
Column Product_Subcategory has 16 values
Column Payment_mode has 4 values
Column Store_type has 4 values
Column city has 10 values
Column Gender has 2 values
Column Education_Level has 5 values
Column Profession_Type has 3 values


### encoding columns to do feature importance using ExtraTreesClassifier and Mutual Information 

In [72]:
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import ExtraTreesClassifier

cat_cols = [
    'Month', 'product_category', 'Product_Subcategory',
    'Payment_mode', 'Store_type', 'city',
    'Gender', 'Education_Level', 'Profession_Type', "Marital_status"
]

num_cols = ['Quantity', 'Unit_Price', 'Total_Price', 'Tax', 'Income', "Reviews", "tax_rate", "price_After_Tax"]

df_encoded = df.copy()
for col in cat_cols:
    le = LabelEncoder()
    df_encoded[col] = le.fit_transform(df_encoded[col].astype(str))


X = df_encoded[cat_cols + num_cols]
y = df_encoded['Return']  

# Model-Based Feature Importance (ExtraTreesClassifier)

In [73]:
# Fit ExtraTreesClassifier
model = ExtraTreesClassifier(n_estimators=100, random_state=42)
model.fit(X, y)

#  Get feature importances
importances = pd.Series(model.feature_importances_, index=X.columns)
importances_sorted = importances.sort_values(ascending=False)

#  Show top features
importances_sorted

tax_rate               0.342999
Reviews                0.242693
Quantity               0.074939
Tax                    0.057281
Total_Price            0.038383
price_After_Tax        0.033160
Product_Subcategory    0.023780
Income                 0.023193
Unit_Price             0.022934
Month                  0.022657
Payment_mode           0.020182
city                   0.017684
product_category       0.016323
Education_Level        0.016206
Store_type             0.014448
Profession_Type        0.013318
Marital_status         0.009951
Gender                 0.009867
dtype: float64

In [ ]:
# Drop low columns importances
df.drop(["Gender", "Marital_status", "Profession_Type", "Education_Level" ], inplace= True, axis= 1)

# Modling 

In [76]:
df_Cat = df.select_dtypes(include="object")

for col in df_Cat.columns :
    print(f"Column {col} has {df_Cat[col].nunique()} values")

Column Month has 12 values
Column product_category has 6 values
Column Product_Subcategory has 16 values
Column Payment_mode has 4 values
Column Store_type has 4 values
Column city has 10 values


In [77]:
df.describe()

Quantity  Unit_Price       Tax   Reviews     Income    Return  \
count 19504.000   19504.000 19504.000 19504.000  19504.000 19504.000   
mean      3.098     776.168   241.235     3.883  70516.888     0.105   
std       1.446     414.249   182.887     1.415  37418.385     0.307   
min       1.000      70.000     7.350     1.000   7157.000     0.000   
25%       2.000     417.000    95.760     3.000  37950.000     0.000   
50%       3.000     773.000   190.785     4.000  69293.000     0.000   
75%       4.000    1135.000   349.965     5.000  99534.000     0.000   
max       5.000    1500.000   787.500     5.000 159984.000     1.000   

       Total_Price  tax_rate  price_After_Tax  
count    19504.000 19504.000        19504.000  
mean      2393.921     0.104         2635.156  
std       1795.395     0.026         1968.175  
min         70.000     0.021           77.350  
25%        945.000     0.105         1046.159  
50%       1930.000     0.105         2126.020  
75%       3522.000     0.105         3875.235  
max       7500.000     0.525         8287.500

In [78]:
Binary_encoding = [ "Month", "Product_Subcategory", "city" ]
    
OneHot_encoding = ["product_category", "Store_type", "Payment_mode" ]

num_columns = ['Unit_Price', 'Total_Price', 'Tax', 'Income', "tax_rate", "price_After_Tax"]

In [79]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder , RobustScaler , StandardScaler , OrdinalEncoder
from category_encoders.binary import BinaryEncoder
from sklearn.model_selection import cross_validate
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV, KFold, cross_validate
from sklearn.metrics import make_scorer, recall_score, accuracy_score
from sklearn.model_selection import cross_validate, StratifiedKFold


In [80]:
Transformation = ColumnTransformer(
    transformers=[
        ("OHE", OneHotEncoder(sparse_output=False, drop="first"), OneHot_encoding),
        ("BE", BinaryEncoder(), Binary_encoding), 
        ("num", RobustScaler(), num_columns)
    ],
    remainder="passthrough"
)

## for test pipline work or not 

In [81]:
steps = [
    ("Transform", Transformation),
    ("LR", LogisticRegression())
]

pipeline = Pipeline(steps=steps)

In [82]:
df.columns

Index(['Month', 'product_category', 'Product_Subcategory', 'Quantity',
       'Unit_Price', 'Tax', 'Payment_mode', 'Store_type', 'Reviews', 'Income',
       'city', 'Return', 'Total_Price', 'tax_rate', 'price_After_Tax'],
      dtype='object')

In [83]:
x = df.drop("Return", axis = 1 )
y = df["Return"] 

In [84]:
kf = KFold(n_splits= 5, shuffle= True, random_state= 20 )
result = cross_validate(pipeline , x ,y , cv = kf , scoring="accuracy" , return_train_score= True)

In [85]:
result["train_score"].mean()

np.float64(0.9461392366933031)

In [86]:
result["test_score"].mean()

np.float64(0.9456522127791033)

In [87]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB , MultinomialNB
# from xgboost import XGBClassifier
from sklearn.neighbors import KNeighborsClassifier


## try all models and chose the best one to do tunung and deploy

In [88]:
models = [
    # ("LR", LogisticRegression()),
    # ("DT", DecisionTreeClassifier()),
      ("RF", RandomForestClassifier()),
    # ("SVM", SVC()),
    # ("GNB", GaussianNB())
    # ("XGB", XGBClassifier())  
]

In [89]:
y.value_counts()

Return
0    17457
1     2047
Name: count, dtype: int64

In [110]:
from sklearn.metrics import make_scorer, recall_score, accuracy_score, precision_score, f1_score
from imblearn.combine import SMOTETomek
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import cross_validate, KFold
from imblearn.pipeline import Pipeline

scoring = {
    'recall': make_scorer(recall_score),
    'accuracy': make_scorer(accuracy_score),
    'precision': make_scorer(precision_score),
}


for model in models:
    # pipeline with EncoTransformation, ScaSMOTETomekler, and the models
    steps = [
       ("Transform", Transformation),
       ("SmoteTomek", SMOTETomek(smote=SMOTE(sampling_strategy={1: 5000}, random_state=24))),
       (model)
]

    pipeline = Pipeline(steps=steps)  # Apply pipeline steps
    
    # Cross-validation
    kf = KFold(n_splits=5, shuffle=True, random_state=21)
    result = cross_validate(pipeline, x, y, cv=kf, scoring=scoring, return_train_score=True)
    
    # Average recall
    print(f"{model[0]} Average Train Recall: {result['train_recall'].mean():.4f}")
    print(f"{model[0]} Average Test Recall: {result['test_recall'].mean():.4f}")
    
    # Average accuracy
    print(f"{model[0]} Average Train Accuracy: {result['train_accuracy'].mean():.4f}")
    print(f"{model[0]} Average Test Accuracy: {result['test_accuracy'].mean():.4f}")
    
    # Average precision
    print(f"{model[0]} Average Train Precision: {result['train_precision'].mean():.4f}")
    print(f"{model[0]} Average Test Precision: {result['test_precision'].mean():.4f}\n")

RF Average Train Recall: 0.9954
RF Average Test Recall: 0.8202
RF Average Train Accuracy: 0.9995
RF Average Test Accuracy: 0.9775
RF Average Train Precision: 0.9995
RF Average Test Precision: 0.9604



## tuning

In [91]:
steps = [
       ("Transform", Transformation),
       ("SmoteTomek", SMOTETomek(smote=SMOTE(sampling_strategy={1: 7000}, random_state=24))),
       ("RF", RandomForestClassifier() )
]

Final_pipeline= Pipeline(steps = steps)

In [92]:
param_grid= { 
                'RF__max_depth': [5, 10, 15],  
                'RF__min_samples_split': [5, 10, 15],  
                'RF__min_samples_leaf': [1, 2, 3]                
} 

In [93]:
from sklearn.model_selection import GridSearchCV, KFold
from sklearn.metrics import make_scorer, recall_score, accuracy_score

scorers = {
    'recall'  : make_scorer(recall_score),
    'accuracy': make_scorer(accuracy_score)
}

kf = KFold(n_splits=5, shuffle=True, random_state=40)
grid = GridSearchCV( Final_pipeline, param_grid= param_grid, cv= kf, scoring= scorers, refit='recall',  return_train_score= True )

result = grid.fit(x, y) 

In [94]:
result.best_params_

{'RF__max_depth': 10, 'RF__min_samples_leaf': 3, 'RF__min_samples_split': 15}

In [95]:
result.cv_results_["mean_train_accuracy"][result.best_index_]

np.float64(0.9834777499505769)

In [96]:
result.cv_results_["mean_test_accuracy"][result.best_index_]

np.float64(0.9787223262937182)

In [97]:
result.cv_results_["mean_train_recall"][result.best_index_]

np.float64(0.8711640063364732)

In [98]:
result.cv_results_["mean_test_recall"][result.best_index_]

np.float64(0.8417826762391991)

## fit final model 

In [99]:
# final model with best estimators
Final_Model = result.best_estimator_

In [100]:
# trains the entire pipeline, including all preprocessing steps and model with best parameters 
Final_Model.fit(x, y)

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/compose/_column_transformer.py:1623: FutureWarning:


The format of the columns of the 'remainder' transformer in ColumnTransformer.transformers_ will change in version 1.7 to match the format of the other transformers.
At the moment the remainder columns are stored as indices (of type int). With the same ColumnTransformer configuration, in the future they will be stored as column names (of type str).
To use the new behavior now and suppress this warning, use ColumnTransformer(force_int_remainder_cols=False).




Pipeline(steps=[('Transform',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('OHE',
                                                  OneHotEncoder(drop='first',
                                                                sparse_output=False),
                                                  ['product_category',
                                                   'Store_type',
                                                   'Payment_mode']),
                                                 ('BE', BinaryEncoder(),
                                                  ['Month',
                                                   'Product_Subcategory',
                                                   'city']),
                                                 ('num', RobustScaler(),
                                                  ['Unit_Price', 'Total_Price',
                                                   'Tax', 'Income', 'tax_rate',
                                                   'price_After_Tax'])])),
                ('SmoteTomek',
                 SMOTETomek(smote=SMOTE(random_state=24,
                                        sampling_strategy={1: 7000}))),
                ('RF',
                 RandomForestClassifier(max_depth=10, min_samples_leaf=3,
                                        min_samples_split=15))])

Quantity  Unit_Price       Tax   Reviews     Income    Return  \
count 19504.000   19504.000 19504.000 19504.000  19504.000 19504.000   
mean      3.098     776.168   241.235     3.883  70516.888     0.105   
std       1.446     414.249   182.887     1.415  37418.385     0.307   
min       1.000      70.000     7.350     1.000   7157.000     0.000   
25%       2.000     417.000    95.760     3.000  37950.000     0.000   
50%       3.000     773.000   190.785     4.000  69293.000     0.000   
75%       4.000    1135.000   349.965     5.000  99534.000     0.000   
max       5.000    1500.000   787.500     5.000 159984.000     1.000   

       Total_Price  tax_rate  price_After_Tax  
count    19504.000 19504.000        19504.000  
mean      2393.921     0.104         2635.156  
std       1795.395     0.026         1968.175  
min         70.000     0.021           77.350  
25%        945.000     0.105         1046.159  
50%       1930.000     0.105         2126.020  
75%       3522.000     0.105         3875.235  
max       7500.000     0.525         8287.500

In [ ]:
df_Cat = df.select_dtypes(include="object")  
for col in df_Cat.columns:                   
    print(f"Unique values in {col}: {df_Cat[col].unique()}\n")  

Unique values in Month: ['January' 'February' 'March' 'April' 'May' 'June' 'July' 'August'
 'September' 'October' 'November' 'December']

Unique values in product_category: ['Books' 'Clothing' 'Home and kitchen' 'Footwear' 'Bags' 'Electronics']

Unique values in Product_Subcategory: ['Fiction' 'Women' 'Bath' 'Mens' 'Kitchen' 'Furnishing' 'Mobiles' 'DIY'
 'Kids' 'Non-Fiction' 'Audio and video' 'Computers' 'Cameras' 'Tools'
 'Academic' 'Personal Appliances']

Unique values in Payment_mode: ['Mobile Payments' 'Credit Card' 'Debit Card' 'Cash']

Unique values in Store_type: ['e-Shop' 'TeleShop' 'MBR' 'Flagship store']

Unique values in city: ['Hyderabad' 'Bangalore' 'Kolkata' 'New Delhi' 'Chennai' 'Pune'
 'Ahmedabad' 'Gurgaon' 'Vishakhapatnam' 'Mumbai']



In [101]:
import pickle

file_name = '/Users/mohammedmahmood/Desktop/Data projects/Retailer_Product_Return_Prediction.sav'

with open(file_name, "wb") as file:
    pickle.dump(Final_Model, file)